third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp

```diff
@@ -144,11 +153,19 @@ void Pingponger::appendSlicedLoadAB(int slice) {
 // are at the memory cluster.
 // Also, SchedBarrier with `0` is set here to tell compiler backend not to
 // reorder any instruction across this point.
-void Pingponger::appendClusterBarrier(OpBuilder &builder, Location loc) {
+SmallVector<Operation *> Pingponger::genClusterBarrier(OpBuilder &builder,
+                                                       Location loc) {
   //  MembarAnalysis can recognize gpu::BarrierOp and skip inserting additional
-  //  barrier
-  appendOp(builder.create<gpu::BarrierOp>(loc));
-  appendOp(builder.create<ROCDL::SchedBarrier>(loc, 0));
+  return {builder.create<gpu::BarrierOp>(loc),
+          builder.create<ROCDL::SchedBarrier>(loc, 0)};
```

Comment:
This is very subtle--I think depends on the C++ standard it can be tricky. IIUC before C++17 there is no guarantee of the evaluation order here so it means you can have them inserted in reverted order.. In Triton we ue C++17 but I'd still want to avoid the subtleness here.. What about using two local variable to explicitly creating them then return?

Reply:
Yes I'm happy to do that. I agree with removing undefined behavior. For my understanding what is the concern with a swapped evaluation order. My understanding was the order of the creates shouldn't matter because we insert into the IR later.

Reply:
Yeah later someone may refactor/update the code here like merging the functions without knowing the trickiness and suddenly we may see ops being inserted in the wrong order.